controller/src/main/java/com/epam/esm/controller/GiftCertificateController.java

```diff
@@ -0,0 +1,116 @@
+     *
+     * @param id the id
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @GetMapping("/certificate/{id}")
+    public ResponseEntity<CertificateDTO> find(@PathVariable(name = "id") Long id) throws ServiceException {
+        CertificateDTO certificateDTO = certificateService.find(id);
+        return ResponseEntity.ok(certificateDTO);
+    }
+
+    /**
+     * Add certificate.
+     *
+     * @param certificateDTO the certificate dto
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @PostMapping(value = "/certificate")
+    public ResponseEntity<CertificateDTO> add(@RequestBody CertificateDTO certificateDTO) throws ServiceException {
+        CertificateDTO result = certificateService.add(certificateDTO);
+        return ResponseEntity.ok(result);
+    }
+
+    /**
+     * Update certificate. Mark the fields that are not specified for updating null.
+     *
+     * @param certificateDTO the certificate dto
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @PutMapping(value = "/certificate", consumes = "application/json")
+    public ResponseEntity<CertificateDTO> update(@RequestBody CertificateDTO certificateDTO) throws ServiceException {
+        CertificateDTO result = certificateService.update(certificateDTO);
+        return ResponseEntity.ok(result);
+    }
+
+    /**
+     * Delete certificate by id.
+     *
+     * @param id the id
+     * @return the response entity
+     * @throws ServiceException the service exception
+     */
+    @DeleteMapping("/certificate/{id}")
+    public ResponseEntity<String> delete(@PathVariable(name = "id") Long id)
+            throws ServiceException, NotExistEntityException {
+        certificateService.delete(id);
+        return ResponseEntity.ok("Delete successful!");
+    }
+
+    /**
+     * Search by tag name or description part
+     *
+     * @param tagName
+     * @param part
+     * @return
+     * @throws ServiceException
+     */
+    @GetMapping("/certificates/search")
+    public ResponseEntity<List<CertificateDTO>> searchByParameter
+    (@QueryParam("tagName") String tagName, @QueryParam("part") String part)
+            throws ServiceException {
+        List<CertificateDTO> certificates = certificateService.searchByParameter(tagName, part);
+        return ResponseEntity.ok(certificates);
+    }
+
+    /**
+     * Sort by date or name ASC/DESC
+     *
+     * @param by
+     * @param type
+     * @return
+     * @throws ServiceException
+     */
+    @GetMapping("/certificates/sort")
+    public ResponseEntity<List<CertificateDTO>> sortByParameter
+    (@QueryParam("by") String by, @QueryParam("type") SortType type)
+            throws ServiceException {
+        List<CertificateDTO> certificates = certificateService.sortByParameter(by, type);
+        return ResponseEntity.ok(certificates);
+    }
+}
```

Comment:
Please do not separate these endpoints, sorting and searching must be done in one endpoint "/certificates". Verbs as "sort" or "search" do not follow Rest API naming best practices (https://restfulapi.net/resource-naming/)

Reply:
rework method